### PR TITLE
✨ 🚧 [WIP]: Feature/add backend

### PR DESCRIPTION
**Pull request by ElRaptorus**
Depends on:
https://github.com/process-engine/process_engine_core/pull/165

https://github.com/process-engine/kpi_api_http/pull/1
https://github.com/process-engine/kpi_api_core/pull/2

https://github.com/process-engine/logging_api_http/pull/1
https://github.com/process-engine/logging_api_core/pull/2
https://github.com/process-engine/logging.repository.file_system/pull/1

https://github.com/process-engine/metrics_api_core/pull/2

https://github.com/process-engine/token_history_api_http/pull/1
https://github.com/process-engine/token_history_api_core/pull/2

**Note that the management API components will only be available as a pre-version, until the KPI-, Logging-, Metrics- and TokenHistory- APIs have been released as packages of their own.**

**Changes:**

Implements the backend for the heatmap PoC.
Contains an implementation for the following API services:
- TokenHistoryApiService
- MetricsApiService
- LoggingApiService
- KpiApiService (WIP)

Contains also config files for using the backend with either a postgres-based database, or with sqlite.

**TODO**:
- Use the metrics service to actually record some metrics in the process engine package (In Review: https://github.com/process-engine/process_engine_core/pull/165 ) 👁‍🗨 
- Finish implementing the KPI service 🚧 
- Use the KPI API to read and evaluate the stored metrics 🚧 

## How can others test the changes?

Startup the backend by navigating to `backend/_application` and then running `npm start`.
Then use the http client of your choice to make http requests against the backend.

## PR-Checklist

Please check the boxes in this list after submitting your PR:

- [x] You can merge this PR **right now** (if not, please prefix the title with "WIP: ")
- [x] I've tested **all** changes included in this PR.
- [x] I've also reviewed this PR myself before submitting (e.g. for scrambled letters, typos, etc.).
- [x] I've rebased the `develop` branch with my branch before finishing this PR.
- [x] I've **summarized all changes** in a list above.
- [x] I've mentioned all **PRs, which relate to this one**.
- [x] I've prefixed my Pull Request title is according to [gitmoji guide](https://gitmoji.carloscuesta.me/).

## Example

<details>
<summary>
:warning: Before merging please click here to expand and follow the guide.
</summary>
<br>

Please use `:twisted_rightwards_arrows:` at the beginning of your merge commit title.

Example 1:

<pre>
<code>
:twisted_rightwards_arrows: :bug: Fix Wrong Text Decoration at ...
</code>
</pre>

To get your commit message, just copy the first part of this pull request.

Example 2:
<pre>
<code>
**Changes:**

- Fixes Wrong Text Decoration at ...
- Fixes some typos
- ...

**Issues:**

Closes #NumberOfFixedIssue

PR: #NumberOfThisPR
</code>
</pre>
</details>

## Emojis

<details>
<summary>
Expand for a list of most used Emojis.
</summary>
<br>

Please prefix your commit messages with an Emoji.

Ref: https://gitmoji.carloscuesta.me/

| Description              | Glyphe               | Emoji  |
|--------------------------|----------------------|--------|
| Bugfix                   | `:bug:`              | 🐛     |
| Configuration releated   | `:wrench:`           | 🔧     |
| Cosmetic                 | `:lipstick:`         | 💄     |
| Dependencies Downgrade   | `:arrow_down:`       | ⬇️     |
| Dependencies Upgrade     | `:arrow_up:`         | ⬆️     |
| Formatting               | `:art:`              | 🎨     |
| Improving Performance    | `:zap:`              | ⚡️     |
| Initial commit           | `:tada:`             | 🎉     |
| Linter                   | `:rotating_light:`   | 🚨     |
| Miscellaneous            | `:package:`          | 📦     |
| New Feature              | `:sparkles:`         | ✨     |
| Refactoring Code         | `:recycle:`          | ♻️     |
| Releasing / Version tags | `:bookmark:`         | 🔖     |
| Removing Stuff           | `:fire:`             | 🔥     |
| Tests                    | `:white_check_mark:` | ✅     |
| Work In Progress (WIP)   | `:construction:`     | 🚧     |

</details>

### CODE REVIEW

File: PULL_REQUEST_TEMPLATE.md [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+## PR-Checklist
+
+Please check the boxes in this list after submitting your PR:
+
+- [ ] You can merge this PR **right now** (if not, please prefix the title with "WIP: ")
+- [ ] I've tested **all** changes included in this PR.
+- [ ] I've also reviewed this PR myself before submitting (e.g. for scrambled letters, typos, etc.)
+- [ ] I've merged the `develop` branch into my branch before finishing this PR.
+- [ ] I've **not added any other changes** than the ones described above.
+- [ ] I've mentioned all **PRs, which relate to this one**
```
Comment (cmg-dev) [inline PULL_REQUEST_TEMPLATE.md]: nimm das aus dem BPMN-Studio

File: backend/metrics_api/src/metrics_service.ts [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+  constructor(loggingRepository: ILoggingRepository) {
+    this._loggingRepository = loggingRepository;
+  }
+
+  private get loggingRepository(): ILoggingRepository {
+    return this._loggingRepository;
+  }
+
+  public async writeOnProcessStarted(correlationId: string, processModelId: string, timestamp: Date): Promise<void> {
+    await this.loggingRepository.writeLogForProcessModel(correlationId, processModelId, LogLevel.info, 'Process started', timestamp);
```
Comment (cmg-dev) [inline backend/metrics_api/src/metrics_service.ts]: bitte alles umbrechen und zwar so:

```ts
    await this.loggingRepository
                .writeLogForProcessModel(
                  correlationId,
                  processModelId,
                  LogLevel.info,
                  'Process started',
                  timestamp);
```

File: backend/metrics_api/src/metrics_service.ts [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+  }
+
+  public async writeOnProcessFinished(correlationId: string, processModelId: string, timestamp: Date): Promise<void> {
+    await this.loggingRepository.writeLogForProcessModel(correlationId, processModelId, LogLevel.info, 'Process finished', timestamp);
+  }
+
+  public async writeOnProcessError(correlationId: string, processModelId: string, timestamp: Date): Promise<void> {
```
Comment (cmg-dev) [inline backend/metrics_api/src/metrics_service.ts]: Nach welcher Regel, brichst du manche signaturen um und manche nicht?

einfach alle umbrechen.

File: backend/metrics_api/src/metrics_service.ts [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+  }
+
+  public async writeOnFlowNodeInstanceResume(correlationId: string,
+                                             processModelId: string,
+                                             flowNodeInstanceId: string,
+                                             flowNodeId: string,
+                                             timestamp: Date): Promise<void> {
+    await this
+      .loggingRepository
+      .writeLogForFlowNode(correlationId, processModelId, flowNodeInstanceId, flowNodeId, LogLevel.info, 'FNI Resumed', timestamp);
```
Comment (cmg-dev) [inline backend/metrics_api/src/metrics_service.ts]: Das kann man mit meinem Lambda sehr schick lösen:

```ts
//Place this at the top of file:

const writeLogForFlowNode: (message: string): Promise<void> = 
  (message: string): Promise<void> => {
    this.loggingRepository
      .writeLogForFlowNode(correlationId, processModelId, flowNodeInstanceId, flowNodeId, LogLevel.info, message, timestamp);
};

// (...)

public async writeOnFlowNodeInstanceSuspend(correlationId: string,
                                              processModelId: string,
                                              flowNodeInstanceId: string,
                                              flowNodeId: string,
                                              timestamp: Date): Promise<void> {

     await writeLogForFlowNode('FNI Suspended');
}

public async writeOnFlowNodeInstanceResume(correlationId: string,
                                             processModelId: string,
                                             flowNodeInstanceId: string,
                                             flowNodeId: string,
                                             timestamp: Date): Promise<void> {

  await writeLogForFlowNode('FNI Exited');
}
```

Damit kann man JEDEN Aufruf in den methoden verschlanken..

File: backend/metrics_api/src/metrics_service.ts (hunk 3 above)
Comment (ElRaptorus) [reply backend/metrics_api/src/metrics_service.ts]: Anhand der in TSLint festgelegten maximalen Zeilenlänge von 150 Zeichen.

File: backend/metrics_api/src/metrics_service.ts (hunk 4 above)
Comment (ElRaptorus) [reply backend/metrics_api/src/metrics_service.ts]: Das klappt so nicht, weil 
a) Die ganzen Parameter wie correlationid, processModelId, etc. and diese Top-Level Function nicht mitgegeben werden und 
b) Sich die `writeOn` Methoden innerhalb einer Klasse befinden, d.h. diese Lambda-Methode hätte keinen Zugriff auf `this.loggingRepository`

Man könnte höchstens sowas hier als private Methode hinzufügen:

```js

  private async _writeLogForFlowNodeInstance(correlationId: string,
                                             processModelId: string,
                                             flowNodeInstanceId: string,
                                             flowNodeId: string,
                                             message: string,
                                             timestamp: Date): Promise<void> {

    await this.loggingRepository
      .writeLogForFlowNode(correlationId, processModelId, flowNodeInstanceId, flowNodeId, LogLevel.info, message, timestamp);
  }
```

Da seh ich aber Null Mehrwert, da die Calls dadurch kein bisschen kleiner werden.

Comment (cmg-dev) [reply backend/metrics_api/src/metrics_service.ts]: Sich das das nicht geht? das müsste über das closure da hinein kommen.

durch die private method gewinne ich halt nix.

File: backend/metrics_api/src/metrics_service.ts (hunk 3 above)
Comment (cmg-dev) [reply backend/metrics_api/src/metrics_service.ts]: ich würde einfach alle umbrechen.

File: backend/logging.repository.file_system/src/logging_repository.ts [hunk 5]
```diff
@@ -0,0 +1,68 @@
+import {ILoggingRepository, LogEntry, LogLevel} from '@process-engine/logging_api_contracts';
+
+import * as path from 'path';
+
+import * as FileSystemAdapter from './adapter';
```
Comment (cmg-dev) [inline backend/logging.repository.file_system/src/logging_repository.ts]: Reihenfolge der imports:

1. System/Framework
1. third party
1. our stuff

:)

File: backend/metrics_api/src/metrics_service.ts (hunk 4 above)
Comment (ElRaptorus) [reply backend/metrics_api/src/metrics_service.ts]: Yep, VSCode hat da sofort gemeckert.

File: backend/logging.repository.file_system/src/logging_repository.ts (hunk 5 above)
Comment (ElRaptorus) [reply backend/logging.repository.file_system/src/logging_repository.ts]: oops :3

**Pull request closed without merging**